src/local/object_id/raw.rs

```diff
@@ -0,0 +1,260 @@
+/// See method documentation for more information.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct RawObjectId {
+    packed: [u32; 3],
+}
+
+impl fmt::Debug for RawObjectId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RawObjectId")
+            .field("packed", &self.packed)
+            .field("real", &self.to_array_string())
+            .finish()
+    }
+}
+
+impl fmt::Display for RawObjectId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut ints = self.non_zero_packed_ints().iter();
+        if let Some(first) = ints.next() {
+            write!(f, "{:x}", first)?;
+        }
+        for next in ints {
+            write!(f, "{:08x}", next)?;
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for RawObjectId {
+    type Err = RawObjectIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, RawObjectIdParseError> {
+        let u128_val = u128::from_str_radix(s, 16)?;
+
+        if u128_val > MAX_PACKED_VAL {
+            return Err(RawObjectIdParseError::value_too_large(u128_val));
+        }
+
+        // if the endianness is right, then I think this should optimize down to a
+        // transmute. If it isn't, then it should be pretty efficient anyways and will
+        // still be _correct_.
+        let as_array = [
+            ((u128_val >> 64) & 0xFFFF_FFFF) as u32,
+            ((u128_val >> 32) & 0xFFFF_FFFF) as u32,
+            (u128_val & 0xFFFF_FFFF) as u32,
+        ];
+
+        Ok(RawObjectId::from_packed(as_array))
+    }
+}
+
+impl RawObjectId {
+    /// Creates an object ID from its packed representation.
+    ///
+    /// The input to this function is the bytes representing the up-to-24 hex
+    /// digits in the object id.
+    pub fn from_packed(packed: [u32; 3]) -> Self {
+        RawObjectId { packed }
+    }
+
+    /// Creates an object ID from a packed representation stored in JavaScript.
+    ///
+    /// The input must be a reference to a length-3 array of integers.
+    ///
+    /// Recommended to be used with the `object_id_to_packed` JavaScript utility
+    /// function, which takes in a string and returns the array of three
+    /// integers that this function expects.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use screeps::{prelude::*, traits::TryInto, RawObjectId};
+    /// use stdweb::js;
+    ///
+    /// let packed_obj_id = (js! {
+    ///     let creep = _.sample(Game.creeps);
+    ///     return object_id_to_packed(creep.id);
+    /// })
+    /// .try_into()
+    /// .unwrap();
+    ///
+    /// let parsed = RawObjectId::from_packed_js_val(packed_obj_id).unwrap();
+    /// println!("found creep with id {}", parsed);
+    /// ```
+    pub fn from_packed_js_val(packed_val: Reference) -> Result<Self, ConversionError> {
+        let mut packed = [0u32; 3];
+        // TODO: make this more efficient, once we get mutable UnsafeTypedArrays.
```

Comment:
Do you mind switching this to a tracking issue? We might never see it again otherwise :P 
(Feel free to leave the todo)

Reply:
Will do!

Reply:
Opened https://github.com/rustyscreeps/screeps-game-api/issues/230.